File: challenges/neural_networks/java/libraries/neural_network_lib/NeuralNetwork.scala
package libraries.neural_network_lib

import matrix_math_lib.Matrix
import matrix_math_lib.MatrixMath

class NeuralNetwork(val input_nodes: Int, val hidden_nodes: Int, val output_nodes: Int) {
  var weights_ih = new Matrix(hidden_nodes, input_nodes)
  var weights_ho = new Matrix(output_nodes, hidden_nodes)
  weights_ih.randomize()
  weights_ho.randomize()

  var bias_h = new Matrix(hidden_nodes, 1)
  var bias_o = new Matrix(output_nodes, 1)
  bias_h.randomize()
  bias_o.randomize()

  var learningRate = 0.1

  def dsigmoid(y: Double): Double = {
    y * (1 - y)
  }

  def setLearningRate(lr: Double): Unit = {
    if (lr < 1 && lr > 0) learningRate = lr
    else System.err.println("Learning rate is out of bounds. Value must be from 0.0 to 1.0")
  }

  def predict(inputs_array: Array[Double]): Array[Double] = {
    val inputs = MatrixMath.fromArray(inputs_array)
    val hidden = MatrixMath.makeMatrixProduct(weights_ih, inputs)
    hidden.add(bias_h)
    hidden.map(ActivationFunction.sigmoid)

    //Generating the hidden output's output
    val outputs = MatrixMath.makeMatrixProduct(weights_ho, hidden)
    outputs.add(bias_o)
    outputs.map(ActivationFunction.sigmoid)

    outputs.toArray();
  }

  def train(inputs_array: Array[Double], targets_array: Array[Double]): Unit = {
    //Generating the hidden outputs
    val inputs = MatrixMath.fromArray(inputs_array)
    val hidden = MatrixMath.makeMatrixProduct(weights_ih, inputs)
    hidden.add(bias_h)
    hidden.map(ActivationFunction.sigmoid)

    //Generating the hidden output's output
    val outputs = MatrixMath.makeMatrixProduct(weights_ho, hidden)
    outputs.add(bias_o)
    outputs.map(ActivationFunction.sigmoid)

    //Convert array of targets to Matrix object
    val convertedTargets = MatrixMath.fromArray(targets_array)

    //Caclculate the error
    val outputErrors = MatrixMath.substract(convertedTargets, outputs)

    //Calculate gradient
    val gradients = MatrixMath.map(outputs, dsigmoid)
    gradients.multiply(outputErrors)
    gradients.multiply(learningRate)

    //Calculate deltas
    val hiddenT = MatrixMath.T(hidden)
    val weightsHoDeltas = MatrixMath.makeMatrixProduct(gradients, hiddenT)

    //Adjust the weights by deltas
    weights_ho.add(weightsHoDeltas)
    //Adjust the bias by its deltas(which are the gradients)
    bias_o.add(gradients)

    //Calculate the hidden layer errors
    val whoT = MatrixMath.T(weights_ho)
    val hiddenErrors = MatrixMath.makeMatrixProduct(whoT, outputErrors)
    val hiddenGradients = MatrixMath.map(hidden, dsigmoid)

    //Calculate hidden gradient
    hiddenGradients.multiply(hiddenErrors)
    hiddenGradients.multiply(learningRate)

    //Calculate input->hidden deltas
    val inputsT = MatrixMath.T(inputs)
    val weightsIhDeltas = MatrixMath.makeMatrixProduct(hiddenGradients, inputsT)

    //Adjust the weights by deltas
    weights_ih.add(weightsIhDeltas)
    //Adjust the bias by its deltas(which are the gradients)
    bias_h.add(hiddenGradients)
  }
}

File: challenges/neural_networks/cpp/digits_recognition/src/neural_network.cpp
#include "neural_network.hpp"
#include <cmath>    
#include <vector>
#include <iostream>

double sigmoid(double x) {
  return 1 / (1 + exp(-x));
}

double dsigmoid(double y) {
  return y * (1 - y);
}

NeuralNetwork::NeuralNetwork(std::vector<int> nodesPerLayer) {
size = nodesPerLayer.size() - 1;
layers = new LayerPtr[size];
for (int i = 0; i < size; i++) {
  layers[i] = new Layer(nodesPerLayer[i], nodesPerLayer[i + 1]);
}
}

NeuralNetwork::~NeuralNetwork() {
  delete layers;
}

Matrix *NeuralNetwork::guess(double inputs_array[], int inputs_size) {
  Matrix *inputs = Matrix::fromArray(inputs_array, inputs_size);
  for (int i = 0; i < size; i++) {
    Matrix *layer_outputs = Matrix::multiply(layers[i]->getInputsWeights(), inputs);
    layer_outputs->add(layers[i]->getInputsBias());
    layer_outputs->map(sigmoid);
    layers[i]->setOutputs(layer_outputs);
    inputs = layer_outputs;
  }
  return inputs;
}

void NeuralNetwork::train(double inputs_array[], double targets_array[], int inputs_size, int targets_size) {
  Matrix *inputs = Matrix::fromArray(inputs_array, inputs_size);
  Matrix *outputs = guess(inputs_array, inputs_size);
  Matrix *convertedTargets = Matrix::fromArray(targets_array, targets_size);
  Matrix *outputErrors = Matrix::substract(convertedTargets, outputs);
  Matrix *gradients;
  for  (int i = size - 1; i >= 0; i--) {
    gradients = Matrix::map(outputs, dsigmoid);
    gradients->multiply(outputErrors);
    gradients->multiply(learningRate);
    Matrix *transposed = (i > 0) ? Matrix::transpose(layers[i - 1]->getOutputs()) : Matrix::transpose(inputs);
    Matrix *weightDeltas = Matrix::multiply(gradients, transposed);
    layers[i]->adjustWeights(weightDeltas);
    layers[i]->adjustBias(gradients);
    if (i > 0) {
      Matrix *wT = Matrix::transpose(layers[i]->getInputsWeights());
      outputErrors = Matrix::multiply(wT, outputErrors);
      outputs = layers[i - 1]->getOutputs();
    }
  }
}